types/basetypes/dynamic_type.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package basetypes
+
+import (
+	"context"
+	"errors"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// DynamicTypable extends attr.Type for dynamic types. Implement this interface to create a custom DynamicType type.
+type DynamicTypable interface {
+	attr.TypeWithDynamicValue
+
+	// ValueFromDynamic should convert the DynamicValue to a DynamicValuable type.
+	ValueFromDynamic(context.Context, DynamicValue) (DynamicValuable, diag.Diagnostics)
+}
+
+var _ DynamicTypable = DynamicType{}
+
+// TODO: doc
+type DynamicType struct{}
+
+// ApplyTerraform5AttributePathStep applies the given AttributePathStep to the type.
+func (t DynamicType) ApplyTerraform5AttributePathStep(step tftypes.AttributePathStep) (interface{}, error) {
+	// MAINTAINER NOTE: Based on dynamic type alone, there is no alternative type information to return related to a path step.
+	// However, it is possible for a dynamic type to have an underlying value of a list, set, map, object, or tuple
+	// that will have corresponding path steps.
+	//
+	// Since the dynamic type has no reference to the underlying value, we just return the dynamic type which can be used
+	// to grab the attr.Value from `(DynamicType).ValueFromTerraform`.
+	return t, nil
+}
+
+// Equal returns true if the given type is equivalent.
+func (t DynamicType) Equal(o attr.Type) bool {
```

Comment:
Perhaps this has already been covered by maintainer note for `ApplyTerraform5AttributePathStep` but is it worth describing why the underlying type (e.g., list) of a dynamic type can't be compared in an equality check?

Reply:
I've added a little blurb to this equal function 👍🏻 

There will probably be more docs coming to the actual `DynamicType` about not having type information about underlying values as well.